scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala

```diff
@@ -8,6 +8,71 @@ case class TypeName(asString: String) extends AnyVal
 sealed trait Projection {
   def andThen(accessor: Accessor, typeName: TypeName): Projection =
     Property(this, accessor, typeName)
+
+  /**
+   * Given a base projection, returns the projection based on it if applicable.
+   *
+   * For instance, given a quoted function
+   * 	 `val contact = Quoted.function { (c: Contact) => c.contact }`
+   * and a call
+   *   `(p: Person) => contact(p.name)`
+   * produces the projection
+   *   `Person.name.contact`
+   */
+  def basedOn(base: Projection): Option[Projection] =
+    this match {
+      case TypeReference(tpe) =>
+        base match {
+          case TypeReference(`tpe`)  => Some(base)
+          case Property(_, _, `tpe`) => Some(base)
+          case other                 => None
+        }
+      case Property(path, name, tpe) =>
+        path.basedOn(base).map(Property(_, name, tpe))
+    }
+
+  /**
+   * Limits projections to only values of `superClass`. Example:
+   *
+   * case class Person(name: String, contact: Contact) extends ThriftObject
+   * case class Contact(phone: Phone) extends ThriftObject
+   * case class Phone(number: String)
+   *
+   * For the super class `ThriftObject`, it produces the transformations:
+   *
+   * Person.contact.phone        => Some(Person.contact.phone)
+   * Person.contact.phone.number => Some(Person.contact.phone)
+   * Person.name.isEmpty         => Some(Person.name)
+   * Phone.number								 => None
+   */
+  def bySuperClass(superClass: Class[_]): Option[Projection] = {
+
+    def isSubclass(c: TypeName) =
+      try
+        superClass.isAssignableFrom(Class.forName(c.asString))
+      catch {
+        case _: ClassNotFoundException =>
+          false
+      }
+
+    def loop(p: Projection): Either[Projection, Option[Projection]] =
```

Comment:
I still find this logic complicated. But comment makes everything better in this case!